ROFO-35 약관 조회 API 개발

Unagi-zoso opened the pull request:
### 개요
회원가입 기능에 사용될 약관 조회 API 를 작성합니다.

### 변경사항
- 약관 Entity 클래스 및 DDL 작성
- 약관 조회 API , Swagger spec 개발
- 유저동의약관 Entity 클래스 및 DDL 작성
- 유저동의약관 저장 기능 작성
- 관련 테스트 코드 및 Fixture 작성
- 관련 커스텀 예외 작성
- 컨트롤러 테스트 시 공통적으로 사용되는 어노테이션을 하나로 관리할 어노테이션 작성
- MockMvc 테스트 중 인증 편의를 위한 메소드 작성
- 피드백 반영 (사용되지 않는 API 제거 및 반환형식 변경, Test Fixture 변경)

### 테스트
- 테스트 코드 추가여부 O

### 관련 지라 및 위키 링크
 - [JIRA](https://weit-2nd.atlassian.net/browse/ROFO-35) 


### 리뷰어에게 하고 싶은 말
모든 피드백 감사히 받습니다!
코드를 작성하고 보니 다른 분들과 충돌이 날만한 기능들이 좀 보입니다.
이런 부분도 기존 기능을 고수하는 게 득이 될 것 같다 싶으면 편하게 말씀해주세요.

Unagi-zoso committed:
ROFO-35 약관 Entity 클래스 및 DDL 작성

Unagi-zoso committed:
ROFO-35 존재하지 않는 약관 접근 시 커스텀 예외 작성 TermNotFoundException.kt

Unagi-zoso committed:
ROFO-35 TermRepository 및 테스트 코드 작성

Unagi-zoso committed:
ROFO-35 약관 DTO 및 TestFixture 작성

Unagi-zoso committed:
ROFO-35 필수약관 미동의 시 커스텀 예외 작성

Unagi-zoso committed:
ROFO-35 TermQueryService 및 테스트 코드 작성

`calculateEachTypeSizes` 메소드는 입력 받은 약관 콜렉션을 사용해 전체 약관 사이즈, 필수 약관 사이즈, 선택 약관 사이즈 셋을 반환하는 기능입니다.
반환타입으로 이 의미를 담을 수 있는 클래스를 새로 만들까 생각도 하였지만 해당 메소드의 사용처가 이후에도 나올 것 같지 않아. `Triple` 로 뒀습니다.

메소드 자체에도 세 가지의 사이즈 반환이라는 역할이 있는데 이걸 하나의 책임으로 봐도 될지 아니면 따로 메소드로 다 분리할 지 고민하였지만
이 부분도 이 메소드의 이후 사용처를 고려하였을 때 이 상태로 관리하여도 괜찮을 것 같다 생각하여 유지하였습니다.

Unagi-zoso committed:
ROFO-35 TermCommandService 및 테스트 코드 작성

`checkRequiredTermsOrThrow` 메소드가 조회의 성향이 강한지 데이터 조작의 성향이 강한지 고민을 했습니다.
저는 조회를 **조회한 무언가를 반환받길 기대하는 행위** 라 생각합니다.
이 메소드는 반환하기 보다는 사용자가 전달한 데이터를 조작해 예외 여부를 검사하는 목적이 강해
`조회` 의 성향과는 거리가 있다 판단해 `TermCommandService` 에 작성하였습니다.

Unagi-zoso committed:
ROFO-35 약관 API 요청 경로를 security 영향을 받지 않도록 설정

Intellij 를 업데이트하면서 플러그인들도 같이 업데이트했는데 그게 영향이 있는지 Ktlint 이 기존 포매팅에 변화를 주었습니다.

Unagi-zoso committed:
ROFO-35 약관 스웨거 태그 추가 및 컨트롤러 테스트를 위한 어노테이션 추가

Unagi-zoso committed:
ROFO-35 TermController 클래스 및 Spec, 테스트 코드 작성

간략한 정보 반환 API 의 요청 경로를 /api/v1/terms/summary 로 하였는데 이게 좀 모호하게 느낄 수도 있을 것 같습니다.
현재는 전체 약관의 수와 전체 약관의 제목을 반환하는데 프론트 분들과 상의해 반환 정보 중 제목만 필요하다하면
/api/v1/terms/title 같은 식으로 정하는 것도 어떨까 생각이 들었습니다.

Unagi-zoso committed:
ROFO-35 유저동의약관 Entity 클래스 및 DDL 작성

조회에 대한 기능은 계획에 없어 인덱스 설정 같은 부분은 추후에 일어날 수 있습니다.

Unagi-zoso committed:
ROFO-35 UserAgreedTermRepository 작성

Hibernate 에서 제공해주는 save 를 제외하곤 아직 사용하는 기능이 없어 테스트 코드는 생략하였습니다.

Unagi-zoso committed:
ROFO-35 UserAgreedTermFixture 작성

불변성이 보장된 요소들은 최상위 프로퍼티로
Entity 나 DTO 의 경우 불변성이 보장되어 있어도 이후 Entity 중 불변성을 보장할 수 없는 객체들이
생겨날 수 있으니 일관성을 위해 생성함수로 만들었습니다.

Unagi-zoso committed:
ROFO-35 UserAgreedTermCommandService 및 테스트 코드 작성

Unagi-zoso committed:
ROFO-35 MockMvc 요청 시 인증편의 기능 추가 및 적용

minseonglove commented on src/main/kotlin/kr/weit/roadyfoody/term/exception/TermNotFoundException.kt:
```diff
@@ -0,0 +1,16 @@
+        private const val TERM_NOT_FOUND_MESSAGE_SUFFIX = "약관을 찾을 수 없습니다."
+
+        @JvmStatic
+        fun termNotFoundMessage(termId: Long) = "$TERM_NOT_FOUND_MESSAGE_PREFIX $termId $TERM_NOT_FOUND_MESSAGE_SUFFIX"
```
앞에 get을 붙여서 의미를 좀 더 명확하게 해도 좋겠어요. 이름이 동사로 시작하지 않으니까 뭔가 변수명 같기도 하네용

Unagi-zoso replied:
그렇네요! 감사합니다

minseonglove commented on src/main/kotlin/kr/weit/roadyfoody/term/service/TermQueryService.kt:
```diff
@@ -0,0 +1,33 @@
+
+    fun getDetailedTerm(termId: Long): DetailedTermResponse = DetailedTermResponse.from(termRepository.getByTermId(termId))
+
+    private fun calculateEachTypeSizes(terms: List<Term>): Triple<Int, Int, Int> {
+        val requiredSize = terms.count { it.requiredFlag }
+        return Triple(terms.size, requiredSize, terms.size - requiredSize)
+    }
```
Triple 대신 적절한 DataClass를 만들어주는 것도 좋아보여용. 메소드 사용처를 봤을때는 이해가 가긴하는데 calculateEachTypeSizes 메소드만 봤을때는 각 요소가 뭘 뜻하는지 바로 알기가 어렵네요.

Unagi-zoso replied:
오! 그렇네요. 
메소드 명을 지어주면서도 많이 찜찜했었는데 반환 타입으로도 힌트를 줄 수 있네요! 
감사합니다

minseonglove commented on src/main/kotlin/kr/weit/roadyfoody/term/service/dto/DetailedTermsResponse.kt:
```diff
@@ -0,0 +1,49 @@
+    val terms: List<DetailedTermResponse>,
+) {
+    companion object {
+        @JvmStatic
+        fun from(
```
요건 제가 써버를 잘 몰라서 그럴 수 있는데 companion 메소드에 JvmStatic 어노테이션을 넣는 이유가 있나용

Unagi-zoso replied:
기존 Java 환경과 호환을 위해 붙였습니다. 코틀린으로만 구성되는 현재 프로젝트에서 그럴 일은 없지만 관습적으로 추가하였습니다.

minseonglove commented on src/main/kotlin/kr/weit/roadyfoody/useragreedterm/domain/UserAgreedTerm.kt:
```diff
@@ -0,0 +1,35 @@
+    ],
+)
+@SequenceGenerator(name = "USER_AGREED_TERMS_SEQ_GENERATOR", sequenceName = "user_agreed_terms_seq", initialValue = 1, allocationSize = 1)
+class UserAgreedTerm(
+    @Id
+    @GeneratedValue(strategy = GenerationType.SEQUENCE, generator = "USER_AGREED_TERMS_SEQ_GENERATOR")
+    val id: Long = 0L,
+    @ManyToOne
+    @JoinColumn(name = "user_id", nullable = false)
+    val user: User,
+    @ManyToOne
+    @JoinColumn(name = "term_id", nullable = false)
+    val term: Term,
+) : BaseTimeEntity()
```
1번 약관을 유저가 동의했지만 1번 약관이 변경되서 재동의를 받아야하는 경우. 지금의 필드만으로는 유저가 동의한 약관이 이전 버전인지 최신 버전인지 구분하기가 힘들어 보이는데 어떤 정책으로 갈 계획인가용

Unagi-zoso replied:
오.. 감사합니다! 정말 좋은 피드백 같아요. 조금 생각을 해볼게요.

Unagi-zoso replied:
#### 우선 현재 시스템에 변경점이 있습니다.
1. `Term` Entity 클래스는 제목, 내용, 필수여부 를 담당한 변수들을 `var` 로 교체합니다.
2. `UserAgreedTerm` Entity 클래스에는 동의여부를 담당하는 변수를 추가합니다. 그리고 해당 변수는 변경이 가능하므로 이 Entity 클래스는 변경이 가능한 Entity 로 취급합니다. (최근 수정 시간 컬럼 추가)

#### 그리고 앞으로 추가될 기능을 고려했습니다.
##### 약관 변경 시
1. 약관의 내용을 바꿉니다.
2. 해당 약관의 ID 에 해당하는 유저 동의 약관의 동의여부를 전부 `false` 로 변경합니다.

##### 로그인 시
1. 소셜 로그인 인증토큰 확인합니다.
2. 유저가 필수약관에 동의했는지 확인합니다.
2-1. 필수약관에 모두 동의했다면 로그인을 성공하여 서비스 인증토큰을 반환합니다.
2-2. 필수약관 중 동의하지 않은게 있다면 전체 미동의 약관의 ID, 제목, 필수여부, 내용 들과 로그인 시도한 회원의 식별자를 담은 임시 토큰도 반환합니다. (비인가 요청 과 약관 동의 요청에만 접근 가능한 권한의 토큰)

해당 토큰의 관리는 유효기간과 한 번 접근하면 폐기하도록 다뤄질 예정입니다.


현재는 이정도로 구상 중입니다. 

---
추신)
혹시 **약관 변경 기능**을 서비스 애플리케이션 단에서 처리한다면 **어드민 서비스 까지 나오고 난 뒤에 사용**이 될 것 같은데 **약관 변경 기능**은 MVP 버전 까지 끝내고 **Extra Feature** 때 구현해도 괜찮을까요? 대신 이번 작업에 기존에 작성한 Entity 와 테이블에는 약관 변경에 대해 대처가 되도록 변경을 하겠습니다.

Unagi-zoso replied:
혹시 약관정보 전달 받을때 필수약관들, 선택약관들 나눠서 
다른 배열로 전달 받는게 편하실까요?

minseonglove replied:
ㅋㅋ 이런 디테일한 구현은 주열님 말대로 최후반으로 미루는게 맞아요. 그래도 이런거 까지 고려하면서 만드는거랑 고려하지 않는거랑 이후 유지보수에서 차이가 날 것이라 생각해 미리 말해봤습니다.

> 혹시 약관정보 전달 받을때 필수약관들, 선택약관들 나눠서 다른 배열로 전달 받는게 편하실까요?

요건 약관에 같이 딸려오는 requiredFlag 필드로 구분하면 되기 때문에 하나로 주는게 편해용

minseonglove reviewed (commented):
👍 👍

Unagi-zoso committed:
ROFO-35 약관 Entity 클래스 제목, 내용, 필수여부 변경 가능하도록 수정

정책에 따라 약관의 내용이 달라질 수 있어 변경이 가능하도록 하였다.

Unagi-zoso committed:
ROFO-35 커스텀 예외 메시지 제작하는 메소드 이름 변경 및 적용

Unagi-zoso committed:
ROFO-35 calculateEachTypeSizes 메소드 이름 변경 및 반환용 DataClass 작성

기존 `calculateEachTypeSizes` 는  전달받은 약관 콜렉션 전체 사이즈와 필수약관, 선택약관 각각의 사이즈를 `Triple` 에 담아 전달하였다.
메소드 자체에 역할도 너무 많고 전달받은 약관 콜렉션 전체 사이즈는 이 메소드가 아니여도 호출한 로직에서 얻을 수 있으므로
메소드의 역할을 필수약관, 선택약관 둘의 사이즈만 구해서 전달하는 것으로 바꿨다.

이 메소드의 조회작업내용이 그렇게 무거워보이진 않아 메소드명도 get 이 앞에 오도록 하였다.
메소드의 역할이 직관적으로 된 것 같아 반환용 DataClass 를 따로 만들까 사용성을 고려했을 때 새로 만들기보다 Pair 를 이용할까 생각하다가 테스트 코드에서도 몇 번 나올 것이고 클래스를 만드는 편이 부르기에도 편할 것 같아 추가하였다.

Unagi-zoso committed:
ROFO-35 Terms, UserAgreedTerms Entity 에 컬럼 및 인덱스 추가

jifrozen0110 commented on src/main/kotlin/kr/weit/roadyfoody/term/exception/TermNotFoundException.kt:
```diff
@@ -0,0 +1,16 @@
+
+class TermNotFoundException(
+    message: String,
+) : BaseException(ErrorCode.NO_SUCH_ELEMENT, message) {
```
> NO_SUCH_ELEMENT를 써도 한동안은 문제 없을것 같은데
> NO_SUCH_ELEMENT가 너무 범용적인 에러코드라 나중에는 이곳 저곳에서 다 쓸것 같네요 ㅋㅋㅋ...
> 유저가 없다는 에러코드를 하나 새로 추가하는게 어땠을까요??

팀장님이 UserNotFoundException에 달아주신 리뷰인데 
TermNotFoundException도 동의 약관 없다는 에러 코드를 추가하여 ErrorCode.NO_SUCH_ELEMENT를 대체하는것은 어떨까요?

Unagi-zoso replied:
아하 감사합니다

jifrozen0110 commented on src/main/kotlin/kr/weit/roadyfoody/term/exception/TermNotFoundException.kt:
```diff
@@ -0,0 +1,16 @@
+        private const val TERM_NOT_FOUND_MESSAGE_PREFIX = "약관 ID:"
+        private const val TERM_NOT_FOUND_MESSAGE_SUFFIX = "약관을 찾을 수 없습니다."
+
+        @JvmStatic
```
companion object 안에서 @JvmStatic을 따로 선언한 이유가 있을까요?

jifrozen0110 replied:
아! 이부분 위에서 확인하였습니다. 
다만 코드의 통일성을 위해 여쭤보는데 static 함수마다 
@JvmStatic 붙이는것을 컨벤션으로 진행하는것이 될까요?

Unagi-zoso replied:
저는 그러는 편이 좋다 생각해요!

jifrozen0110 replied:
저는 저희 프로젝트가 자바와 호환할 경우가 거의 없다고 생각하여
굳이 싶기는 합니다.
만약 자바와 호환할 경우가 생긴다면
그때는 자바 파일에서 사용하는 함수만 JvmStatic 선언을 해주면 되는 것이 아닐까요?

Unagi-zoso replied:
맞는 말씀이세요. 
컨벤션 관련된 부분이다보니 다른 분들 의견도 들어보고 적용해볼게요. 감사합니다 !

konempty replied:
요건 케바케라고 생각하는데
꼭 Java와의 호환성을 제외하고도 저게 강요되는 케이스가 있고 반대로 대체하는 방법도 있어요
가장 대표적인 필수 케이스의 경우
companion object의 동작 구조상 비용이 비싼 객체의 경우 여러번 생성되는걸 막기위해 JVMStatic을 붙혀두는게 이득인 경우가 있고
대체 하는 방법중에 가장 간단한건 상수를 const val로 선언하면 JvmStatic이 없어도 스태틱하게 선언이 됩니다~!

하지만 지금의 경우처럼 함수를 static화 하는건 필수 사항은 아니고 하든 안하든 상관없으며 갠취입니다!
참고로 저는 이경우 안쓰는걸 더 선호해요 ㅋㅋㅋㅋ

jifrozen0110 commented on src/test/kotlin/kr/weit/roadyfoody/term/fixture/TermFixture.kt:
```diff
@@ -0,0 +1,122 @@
+val TEST_REQUIRED_TERM_3_CONTENT = coverContentHtmlTag("테스트 필수 약관 $TEST_REQUIRED_TERM_3_ID 내용")
+const val TEST_REQUIRED_TERM_3_REQUIRED_FLAG = true
+
+fun createTestRequiredTerm3() =
```
createTestRequiredTerm1, createTestRequiredTerm2, createTestRequiredTerm3
이렇게 계속 중복되는데

`fun createTestRequiredTerms() = (1L..3L).map { createTestRequiredTerm(it) }`

이런식으로 정리할 순 없을까요??

Unagi-zoso replied:
관리하는게 번거롭긴 했는데 한 번 적용해볼게요

jifrozen0110 reviewed (approved):
고생하셨습니다!

Unagi-zoso committed:
ROFO-35 NO_SUCH_ELEMENT 에러코드 대체 커스텀 예외 에러코드 추가

Unagi-zoso committed:
ROFO-35 Term, UserAgreedTerm Test Fixture 생성 방식 변경

현재 방식은 하드코딩 하는 느낌이 있어 관리하기가 복잡했습니다. 현재 방식은 그런 속성들을 패러미터로 주입 받아
코드 중복을 줄였습니다.
Test Fixture 는 기본적으로 컬렉션을 생성하게 하여 단일 데이터가 필요하면 first() 메소드를 통해 꺼내 사용하게 했습니다.

Unagi-zoso committed:
ROFO-35 잘못된 테스트 내용 수정

앞으론 더욱 주의하겠습니다.

konempty commented on src/main/kotlin/kr/weit/roadyfoody/term/domain/Term.kt:
```diff
@@ -0,0 +1,31 @@
+    var title: String,
+    @Column(columnDefinition = "CLOB", nullable = false)
+    var content: String,
+    @Column(name = "required_flag")
```
요 컬럼은 이름을 명시해준 이유가 있을까요??

Unagi-zoso replied:
예전에 공부할 때 이런 것도 명시해주는 편이 좋다라 배워서 쓰고 있었습니다.
이건 변수명 변경하면서 같이 지우겠습니다.

konempty commented on src/main/kotlin/kr/weit/roadyfoody/term/exception/TermNotFoundException.kt:
```diff
@@ -0,0 +1,16 @@
+        private const val TERM_NOT_FOUND_MESSAGE_SUFFIX = "약관을 찾을 수 없습니다."
+
+        @JvmStatic
+        fun getTermNotFoundMessage(termId: Long) = "$TERM_NOT_FOUND_MESSAGE_PREFIX $termId $TERM_NOT_FOUND_MESSAGE_SUFFIX"
```
제가 코드를 다 읽은게 아니라 잘못된 코멘트 일수도 있는데 어차피 요 메소드는 객체 생성시 사용될것 같은데
요 부분을 차라리 long을 받는 생성자로 처리하는게 어땠을까요??

konempty commented on src/main/kotlin/kr/weit/roadyfoody/term/service/TermCommandService.kt:
```diff
@@ -0,0 +1,22 @@
+        val requiredTermIds = termRepository.findAllIdsByRequiredFlagIsTrue()
+        val agreedTermIdSet = agreedTermIds.toSet()
+
+        val notAgreedTermIdSet = requiredTermIds.subtract(agreedTermIdSet)
```
requiredTermIds의 자료형이 List인데 List의 subtract는 복잡도가 O(N^2)입니다
agreedTermIdSet을 Set으로 쓴건 아마 O(N log N)의 복잡도를 노리신게 아닐까 싶은데
그렇다면 requiredTermIds의 자료형도 Set으로 바꾸는건 어떨까요??

konempty replied:
요것도 제가 잘못알았네요 ㅋㅋㅋ
내부적으로 Set으로 만들어서 비교를 돌리네요
그래도 requiredTermIds의 자료형이 Set이면 좋을것 같아요!
<img width="603" alt="스크린샷 2024-06-16 오후 7 07 41" src="https://github.com/Weit-2nd/roady-foody-server/assets/44064530/74f5fc19-52ee-4000-8d7f-c7323791212d">

konempty commented on src/main/kotlin/kr/weit/roadyfoody/term/service/TermCommandService.kt:
```diff
@@ -0,0 +1,22 @@
+) {
+    fun checkRequiredTermsOrThrow(agreedTermIds: List<Long>) {
+        val requiredTermIds = termRepository.findAllIdsByRequiredFlagIsTrue()
+        val agreedTermIdSet = agreedTermIds.toSet()
```
checkRequiredTermsOrThrow의 사용처가 없어서 어떻게 사용하실진 모르겠지만
클라에서 보낸 agreedTermIds를 Set으로 받아들이도록 하다면 별도의 변환과정은 없어도 되지않을까요?

konempty commented on src/main/kotlin/kr/weit/roadyfoody/term/service/dto/DetailedTermsResponse.kt:
```diff
@@ -0,0 +1,49 @@
+import com.fasterxml.jackson.annotation.JsonProperty
+import kr.weit.roadyfoody.term.domain.Term
+
+data class DetailedTermsResponse(
```
클라에서 처음 약관의 개수를 알때는 SummaryTermsResponse만 있으면 될것 같고
반대로 각 약관의 내용을 읽어보고 싶을땐 DetailedTermResponse만 있으면 될것 같은데
DetailedTermsResponse를 클라가 필요로 할때는 어떤 상황일까요??

Unagi-zoso replied:
클라이언트 쪽에서 필요한 상황이 오지 않을까? 만드는데 큰 비용이 드는 기능도 아니니 
일단 추가해보자는 생각으로 만들었네요.
이런 습관이 반복되면 오버엔지니어링을 자주하게 되지 않을까 생각이 드네요.
이런 API 를 만들 땐 만들기전에 어떤 기능이 필요한지 여쭤보고 작업하는게 좋을까요?

konempty replied:
어떤 기능이 필요한지 보다는
최소한의 스펙으로 만들고 클라에서 추가요청을하면 그때 추가하는 방식을 저는 선호합니다~!

Unagi-zoso replied:
아하 기획을 좀 더 잘이해하도록 관심을 가지겠습니다.

konempty commented on src/main/kotlin/kr/weit/roadyfoody/term/service/dto/SummaryTermsResponse.kt:
```diff
@@ -0,0 +1,45 @@
+    val allTermsSize: Int,
+    @JsonProperty("required_terms_size")
+    val requiredTermsSize: Int,
+    @JsonProperty("optional_terms_size")
```
모든 필드에 일일히 이름을 지정해주시는 이유가 있을까요??
심지어 전부 스네이크 케이스를 사용하신것 같은데 카멜 케이스가 싫었다면 ObjectMapper를 수정하는게 맞을것 같아요..!
그리고 클라는 카멜케이스를 더 사용하기 편할걸로 알고 있는데 스네이크 케이스를 써달라는 별도의 요청이 있었던걸까요..?!

Unagi-zoso replied:
외부 API 쓸 때 스네이크 케이스를 자주 봐서 스네이크 케이스를 사용하는 분위기인줄 알았습니다.
클라이언트 분에게 여쭤보고 카멜케이스로 바꿔보겠습니다.

konempty replied:
아 요거 제가 잘못알았나보네요 ㅋㅋㅋ
확인해보니까 스네이크가 국룰이라고 하네요

konempty commented on src/main/kotlin/kr/weit/roadyfoody/useragreedterm/domain/UserAgreedTerm.kt:
```diff
@@ -0,0 +1,43 @@
+    indexes = [
+        Index(name = "user_agreed_terms_agreed_flag_idx", columnList = "agreed_flag"),
+        Index(name = "user_agreed_terms_user_id_idx", columnList = "user_id"),
+        Index(name = "user_agreed_terms_term_id_idx", columnList = "term_id"),
```
ㅋㅋㅋㅋㅋㅋㅋ
저는 오라클에서 이게 불편해요
FK를 걸었는데 또 인덱스까지 추가로 선언해줘야하더라구요

konempty commented on src/main/kotlin/kr/weit/roadyfoody/useragreedterm/domain/UserAgreedTerm.kt:
```diff
@@ -0,0 +1,43 @@
+    @GeneratedValue(strategy = GenerationType.SEQUENCE, generator = "USER_AGREED_TERMS_SEQ_GENERATOR")
+    val id: Long = 0L,
+    @Column(name = "agreed_flag")
+    var agreedFlag: Boolean = true,
```
저희가 이름을 담는 컬럼을 nameString이라고 하지는 않는것 처럼 boolean이라는 점에서 이미 flag로 사용된다는건 알수있을것 같아서 이름 뒤에 flag는 빼도 될것 같아요!
약관 테이블에 있는 requiredFlag도 동일할것 같구요!

konempty commented on src/main/kotlin/kr/weit/roadyfoody/useragreedterm/exception/RequiredTermNotAgreedException.kt:
```diff
@@ -0,0 +1,17 @@
+        private const val REQUIRED_TERM_NOT_AGREED_MESSAGE_SUFFIX = "필수 약관에 동의하지 않았습니다."
+
+        @JvmStatic
+        fun getRequiredTermNotAgreedMessage(termIds: Iterable<Long>) =
```
여기도 생성자로 바꿀수 있을듯하네요

konempty commented on src/main/kotlin/kr/weit/roadyfoody/useragreedterm/service/UserAgreedTermCommandService.kt:
```diff
@@ -0,0 +1,25 @@
+            terms.map { term ->
+                UserAgreedTerm(user = user, term = term)
+            }
+        userAgreedTermRepository.saveAll(userAgreedTerms)
```
테이블 구조만 봤을때에는 동의하지 않는 약관도 저장하는것 같았는데
실제로는 동의하는 컬럼만 저장하는군요..!
그렇다면 약관 동의 테이블에 agreedFlag의 의의는 뭘까요..??

konempty commented on src/main/resources/db/migration/V2__create_terms.sql:
```diff
@@ -0,0 +1,17 @@
+    id               NUMERIC(19, 0)      NOT NULL,
+    title            VARCHAR2(90)        NOT NULL,
+    content          CLOB                NOT NULL,
+    required_flag     NUMBER(1) DEFAULT 0 NOT NULL,
```
ㅋㅋㅋㅋ 얘만 뭔가 인덴트가 이상하네요
코드 정렬 한번부탁드려요!

konempty commented on src/test/kotlin/kr/weit/roadyfoody/term/repository/TermRepositoryTest.kt:
```diff
@@ -0,0 +1,98 @@
+            context("존재하지 않는 id 리스트를 받는 경우") {
+                it("빈 리스트를 반환한다.") {
+                    val terms = termRepository.findAllByIdIn(listOf(TEST_NONEXISTENT_TERM_ID))
+                    terms.size shouldBe 0
```
kotest에서는 많은 확장함수를 지원하는데
`terms.shouldBeEmpty()`
를 사용하는게 더 의도를 잘 설명하는 테스트 검증코드 아닐까요??

konempty commented on src/test/kotlin/kr/weit/roadyfoody/term/repository/TermRepositoryTest.kt:
```diff
@@ -0,0 +1,98 @@
+                    val terms = termRepository.findAllByIdIn(givenTerms.map { it.id })
+                    repeat(givenTerms.size) {
+                        terms[it] shouldBe givenTerms[it]
+                    }
```
kotest에서는 컬렉션끼리의 비교도 잘해주는 편인데
`terms shouldBe givenTerms`
요렇게 반복문 없이 한줄로 처리하면 더 예쁠것 같아요!

konempty commented on src/test/kotlin/kr/weit/roadyfoody/term/service/TermCommandServiceTest.kt:
```diff
@@ -0,0 +1,59 @@
+                    shouldNotThrow<RequiredTermNotAgreedException> {
+                        termCommandService.checkRequiredTermsOrThrow(createTestTermIds())
+                    }
+                    verify { termRepository.findAllIdsByRequiredFlagIsTrue() }
```
요 라인을 어떤 정보를 증명하기 위한 줄일까요??
아무런 정보를 증명하지 않는것 같아서 질문드립니다!

Unagi-zoso replied:
`termCommandService.checkRequiredTermsOrThorw` 메소드 로직 상 전체 필수약관의 ID를 가져오는데 

```
 fun checkRequiredTermsOrThrow(agreedTermIds: List<Long>) {
        val requiredTermIdSet = termRepository.findAllIdsByRequiredIsTrue()
        val notAgreedTermIdSet = requiredTermIdSet.subtract(agreedTermIds.toSet())
```
이게 이 메소드의 테스트들한테 다 적용되는 부분이라 given 아래에 stubbing 로직을 뒀습니다.
```
given("checkRequiredTermsOrThrow 테스트") {
    every { termRepository.findAllIdsByRequiredIsTrue() } returns createTestRequiredTermIds()
```

이렇게 하면 중첩된 스코프에도 stubbing 의 영향이 갈 것이라 생각했습니다. 
제가 혹시 오해하고 있는 부분이 있을까요?

konempty commented on src/test/kotlin/kr/weit/roadyfoody/test/controller/TestControllerTest.kt:
```diff
@@ -1,23 +1,18 @@
 // TODO : TestControllerTest.kt 를 걷어낼 때 같이 걷어내면 됩니다.
-@ActiveProfiles("test")
-@Import(TraceManager::class, SecurityConfig::class, ObjectMapperProvider::class, CustomAuthenticationEntryPoint::class)
+@ControllerTest
```
👍 👍

konempty commented on src/test/kotlin/kr/weit/roadyfoody/useragreedterm/fixture/UserAgreedTermFixture.kt:
```diff
@@ -0,0 +1,44 @@
+fun createTestUser1AgreedRequiredTerm(
+    id: Long,
+    termId: Long,
+) = UserAgreedTerm(
```
이름과는 달리 사용자의 id를 동적으로 넣음으로써 user1이 아닌 다른 유저의 정보도 생성할수 있을것 같은데
뭔가 괴리감이 느껴지네요 ㅠㅠ

Unagi-zoso replied:
네. 이 부분은 UserFixture 에도 변경이 필요해 보여서 
회원가입 기능을 만들면서 User 에 변경을 줄 때 같이 작업을 하면 어떨까 해서 이번 작업에선 방치해뒀습니다. 
이 부분을 PR 에 명시해뒀으면 좋았겠네요. 감사합니다.

konempty reviewed (changes requested):
고생하셨습니다 👍 👍 
확실히 꼼꼼히 작업해주셨고
많은 고민을 해주신것 같아서 보는 제가 뿌듯한 작업이네요 ㅋㅋㅋㅋ
쪼금만 고생부탁드립니다!

konempty commented:
<img width="780" alt="스크린샷 2024-06-16 오후 6 38 04" src="https://github.com/Weit-2nd/roady-foody-server/assets/44064530/2dff3929-38ec-43a2-9fbd-46d76e4e68fe">

아 그리고 한가지 안타까운게 있는데
저번 user테이블의 경우 간단한 테이블이었으니까 괜찮았는데
이번에는 꽤 복잡한 구조였던 만큼 ERD를 미리 그려보고 서버 개발자들의 의견을 받아보셨어야 합니다..!
그 과정이 생략되어서 그런지 Flyway 스크립트가 3개나 나온게 아닐까 생각되구요 ㅋㅋ큐ㅠㅠ

Unagi-zoso commented:
아이고 이건 제가 놓쳐버렸네요.. 앞으론 꼭 피드백을 받고 작업하겠습니다..

Unagi-zoso committed:
ROFO-35 terms , user_agreed_terms Flyway 스크립트 작성

Unagi-zoso committed:
ROFO-35 사용되지 않는 Flyway 스크립트 제거

Unagi-zoso committed:
ROFO-35 커스텀 예외 메시지 생성 방식 생성자로 변경

Unagi-zoso committed:
ROFO-35 Term Entity 변수, 컬렴명 변경 requiredFlag -> required

Unagi-zoso committed:
ROFO-35 Test Fixture 수정

UserAgreedTermFixture 의 경우 user 1 을 고정적으로 생성했었는데 이를 외부에서 주입 받을 수 있게 만들었습니다.
그외 메소드 패러미터, 반환타입의 변경으로 인해 수정하였습니다.

Unagi-zoso committed:
ROFO-35 checkRequiredTermsOrThrow 메소드 패러미터 타입 Set 으로 변경

현재 계획 중인 기능 상 List 로 다뤄질 일이 크게 없어 외부에서 입력 받는 시점부터(Request Dto) Set 으로 받을 예정입니다.
따라서 굳이 내부에서 Set 으로 변경할 이유가 사라졌습니다.

Unagi-zoso committed:
ROFO-35 getAllDetailedTerms 메소드 제거 (미사용)

현재 기획 상 사용되지 않는 기능이라 제거합니다.

Unagi-zoso committed:
ROFO-35 사용되지 않는 API 제거 및 약관 Response 삭제, 컬럼명 수정

클라이언트 작업자분과 의견을 나누어 다음과 같이 변경하였습니다.

스네이크 케이스 -> 카멜 케이스
required_flag -> isRequired

Unagi-zoso committed:
ROFO-35 Test Fixture 변경으로 인한 수정

Kotest 기능을 좀 더 확실히 사용할 수 있게
기존 shouldBe true 같은 부분은 shouldBeTrue() 같이 변경하였습니다.

Unagi-zoso committed:
ROFO-35 UserAgreedTerm Entity 쓰이지 않는 변수, 컬럼 및 인덱스 제거

Unagi-zoso committed:
ROFO-35 storeUserAgreedTerms 메소드 패러미터 타입 변경 List -> Set

Unagi-zoso committed:
ROFO-35 Test Fixture 잘못된 로직 수정

이미 `(TEST_REQUIRED_TERMS_SIZE + 1L..TEST_REQUIRED_TERMS_SIZE + TEST_OPTIONAL_TERMS_SIZE)` 이 부분에서 이전 필수약관의 수를 포함하였는데 `termId` 인자 부분에서 중복적으로 더하여서 제거했습니다.

konempty commented on src/main/kotlin/kr/weit/roadyfoody/term/presentation/api/TermController.kt:
```diff
@@ -0,0 +1,24 @@
+class TermController(
+    private val termQueryService: TermQueryService,
+) : TermControllerSpec {
+    @GetMapping("/summary")
```
저는 summary라는 추가 path는 불필요하다고 생각하는데 어떻게 생각하시나요??
/api/v1/terms로만 끝나도 괜춘할것 같아요!

konempty commented on src/main/kotlin/kr/weit/roadyfoody/term/service/TermCommandService.kt:
```diff
@@ -0,0 +1,19 @@
+) {
+    fun checkRequiredTermsOrThrow(agreedTermIdSet: Set<Long>) {
+        val requiredTermIdSet = termRepository.findAllIdsByRequiredIsTrue()
+        val notAgreedTermIdSet = requiredTermIdSet.subtract(agreedTermIdSet)
```
<img width="577" alt="스크린샷 2024-06-17 오후 11 23 26" src="https://github.com/Weit-2nd/roady-foody-server/assets/44064530/511e3fc4-4b41-4579-969d-f2147c8ee013">

둘다 set일 경우 
`val notAgreedTermIdSet = requiredTermIdSet.minus(agreedTermIdSet)`
를 쓰는게 성능상 미미하지만 이득이 있어요 ㅋㅋㅋ

Unagi-zoso replied:
오.. 조금 생각을 해봤는데 어떤 이득인지 잘 모르겠네요..
`return this.filterNotTo ..`이 방식이 subtract() 내부의 removeAll() 호출 보다 이점이 있는건가요?
그리고 제가 구현한 메소드 시간복잡도가 O(N) 이라 생각하는데 맞을까요?

konempty replied:
agreedTermIdSet의 크기가 N requiredTermIdSet의 크기가 M 이라고 가정할때
removeAll은 아래와 같이 Collection<E> 메소드 인걸로 확인이되는데
<img width="516" alt="스크린샷 2024-06-18 오전 10 31 15" src="https://github.com/Weit-2nd/roady-foody-server/assets/44064530/66a2889d-9493-4a9f-bcb8-d4d24ae8630b">
Collection<E>은 구조상 특정 엘리먼트를 찾기위해 O(N)의 연산이 필요합니다 
(만약 내부적으로 Set을 인식하는 코드가 있다면 in 연산과 동일)
따라서 subtract는 O(N * M)의 연산이 나올걸로 예상되는데
Set<T>.minus에서 사용하는 it in other의 경우 TreeSet일때 O(log N) 혹은 HashSet일때 O(1)의 연산입니다
<img width="519" alt="스크린샷 2024-06-18 오전 10 32 06" src="https://github.com/Weit-2nd/roady-foody-server/assets/44064530/3ad56b2c-fd0f-4316-8394-04a0d1c61173">

따라서 Set<T>.minus 전체 복잡도는 O(M log N)이거나 O(M)일것 같은데 높은확률로 후자일것 같아요!

사실 제가 removeAll를 까볼려고 했는데 내부 구현이 안보이도록 되어있더라구요 ㅋㅋ큐ㅠㅠ
그래서 저는 내부 구현을 예상만 해야하는 subtract과는 달리 빠르다는게 보장된 minus의 사용을 추천드린거예요 ㅋㅋㅋ

konempty replied:
확실한 근거는 없고 제 예상에 근거한 리뷰라 혼란을 드렸던것 같네요 ㅋㅋ큐ㅠㅠㅠ

konempty replied:
사실 요대로 놔둬도 괜찮을것 같아요 ㅋㅋㅋㅋ

Unagi-zoso replied:
이 부분 알려주신거 정말 좋은거 같아요.
코틀린, 기타 라이브러리에서 제공해주는 기능들 쓸 때 세부적인 동작은
신경 안쓰고 사용했는데 이젠 이런 부분도 다 신경쓰는 습관이 생길 거 같아요!

konempty commented on src/main/kotlin/kr/weit/roadyfoody/term/service/TermQueryService.kt:
```diff
@@ -0,0 +1,31 @@
+    fun getAllSummaryTerms(): SummaryTermsResponse {
+        val terms = termRepository.findAll()
+        val (requiredSize, optionalSize) = getRequiredAndOptionalSizes(terms)
+        return SummaryTermsResponse.from(terms.size, requiredSize, optionalSize, termRepository.findAll())
```
위에서 termRepository.findAll()를 하는데 여기서 한번 더하는군요 ㅋㅋㅋ
위에서 가져온 terms를 넘겨주는게 어떨까요??

konempty commented on src/main/kotlin/kr/weit/roadyfoody/term/service/TermQueryService.kt:
```diff
@@ -0,0 +1,31 @@
+    }
+}
+
+data class RequiredAndOptionalSizes(
```
얘는 SummaryTermsResponse를 위해 필요한것 같은데
SummaryTermsResponse에서 알아서 계산하면 RequiredAndOptionalSizes 요 DTO는 없어질수 있지 않을까요??

Unagi-zoso replied:
조금이라도 복잡한 로직은 service 레이어 같은 부분에서 담당을 해야한다고 생각했는데
이 정도 작업은 DTO 생성자한테 맡겨도 괜찮을 것 같네요.

konempty commented on src/main/kotlin/kr/weit/roadyfoody/term/service/dto/DetailedTermResponse.kt:
```diff
@@ -0,0 +1,21 @@
+    val content: String,
+) {
+    companion object {
+        @JvmStatic
```
다른곳은 다 사라진것 같은데 여긴 JvmStatic이 남아있군요 ㅋㅋㅋ

konempty commented on src/main/kotlin/kr/weit/roadyfoody/useragreedterm/exception/RequiredTermNotAgreedException.kt:
```diff
@@ -0,0 +1,14 @@
+import kr.weit.roadyfoody.common.exception.ErrorCode
+
+class RequiredTermNotAgreedException(
+    termIds: Iterable<Long>,
```
여기는 Set밖에 안들어 올것 같은데 Iterable로 추상적으로 쓴 이유가 있을까요??
요 코드 의도만 좋다면 좋은 코드일것 같은데 의도가 궁금하네요 ㅋㅋㅋㅋ

Unagi-zoso replied:
현재 기능상 ID 콜렉션들을 다 Set 으로 다루고 있어서 이것도 명시적으로 Set 으로 두는 편이
명료할까 아니면 이후 Set 말고도 List 같은 자료형으로 ID 를 다룰지도 모르니 Iterable 을 사용하는 편이
나을까 고민을 했습니다. 
이게 오버 엔지니어링 수준의 문제는 아닐 것 같아서 범용성 있게 Iterable 로 뒀습니다. 구체적인 구현객체를 두는게
확장성에 떨어진다고 알고 있어 이 부분도 조금 염려에 뒀습니다.

근데 또 마음에 걸리는게 Repository 에서 ID 를 인자로 받을 땐 Set 으로 명시했어서 찜찜한 구석이 있습니다..

konempty replied:
딱 제가 생각한 의도인것 같고 쪼금 과한 대비일수도 있을것 같은데 나쁘진 않은것 같아요 ㅋㅋㅋ

konempty commented on src/main/kotlin/kr/weit/roadyfoody/useragreedterm/service/UserAgreedTermCommandService.kt:
```diff
@@ -0,0 +1,25 @@
+    private val userAgreedTermRepository: UserAgreedTermRepository,
+    private val termRepository: TermRepository,
+) {
+    fun storeUserAgreedTerms(
```
여기는 트랜잭션이 필요해보이네요~!

Unagi-zoso replied:
제가 아직 트랜잭션 걸어주는 기준을 좀 헷갈려해서 여쭈고 싶네요.
UserAgreedTerm 을 영속화할 때까지 영속성 컨텍스트를 유지하기 위함인가요?
아니면 로직을 원자적으로 다루기 위해선가요? 

혹시 어떤 문제가 일어날 수 있는지 조금 알려주실 수 있을까요?

Unagi-zoso replied:
제가 아직 트랜잭션을 좀 헷갈려해서 여쭈고 싶네요.
UserAgreedTerm 을 영속화할 때까지 영속성 컨텍스트를 유지하기 위함인가요?
아니면 로직을 원자적으로 다루기 위해선가요?

혹시 어떤 문제가 일어날 수 있는지 조금 알려주실 수 있을까요?

konempty replied:
요 케이스의 경우 로직을 원자적으로 처리하기 위함입니다~!
아주아주 낮은 확률이지만 레이스 컨디션의 발생할 가능성이 보여서 리뷰드렸습니다 ㅋㅋㅋ

konempty commented on src/test/kotlin/kr/weit/roadyfoody/term/fixture/TermFixture.kt:
```diff
@@ -0,0 +1,59 @@
+// Fail Case
+const val TEST_NONEXISTENT_TERM_ID = 0L
+
+fun createTestNotEnoughRequiredTermIdSet() = createTestRequiredTerms().run { this.subList(0, this.size - 1) }.map { it.id }.toSet()
```
`createTestRequiredTerms().dropLast(1).map { it.id }.toSet()`
요렇게 하는게 의도를 더 잘나타내는 코드가 아닐까요?? ㅋㅋㅋㅋ

Unagi-zoso replied:
앗 이게 훨씬 좋네요. 감사합니다.

konempty commented on src/test/kotlin/kr/weit/roadyfoody/term/service/TermCommandServiceTest.kt:
```diff
@@ -0,0 +1,59 @@
+                    shouldNotThrow<RequiredTermNotAgreedException> {
+                        termCommandService.checkRequiredTermsOrThrow(createTestTermIdSet())
+                    }
+                    verify { termRepository.findAllIdsByRequiredIsTrue() }
```
아 ㅋㅋㅋ verify의 사용에 대해 제가 헷갈렸네요
저는 주로 몇번 호출될것이라 기대하는지까지 적어두거든요 
verify(exactly = 1)
요런식으로요 ㅋㅋㅋ 그런데 저 숫자가 없어서 헷갈렸네요

여기도 명시적으로 적어두는건 어떨까요??

Unagi-zoso replied:
피드백 너무 좋습니다. 
verify 의 exactly 의 기본값이 1 인줄 알았는데 아닌 것 같네요.
해결방법으로 afterTest { clearAllMocks() } 을 하던가 isolationMode 를 leaf 로 두던가 
둘 중 한 가지를 방법으로 보고 있는데 어떤게 성능적으로 좀 더 나을까요? 전 clearAllMocks() 가 괜찮아보여서 적용했습니다.

제가 파악하고 있는 원인으로는 기본적으로 spec 이 싱글톤으로 만들어져서 재사용되는데 
제가 하던 방식은 Mock 객체를 초기화하지도 않고 중복적으로 사용해 
verify exactly 를 1로 명시했을 때 다 문제가 생기는 것으로 생각하고 있습니다. 맞을까요?

konempty replied:
1. > verify exactly 를 1로 명시했을 때 다 문제가 생기는 것으로 생각하고 있습니다. 맞을까요?

네넵 일단 저도 요 현상이 발생하지 않을까 생각했던것도 있고
그냥 verify라고 하면 2번이상 호출되는것도 OK가 되어버리는데 
사실 한번만 실행되는게 의도였다면 해당 테스트는 잘못된 검증을 진행하게되는거니까
좀더 타이트 하게 잡아도 되지 않을까 싶어서 리뷰 드렸습니다~! ㅋㅋㅋ
근데 쫌 투머치 일수도 있겠네요

2. >  전 clearAllMocks() 가 괜찮아보여서 적용했습니다.

저도 둘중하나를 선택해야한다면 요게 더 나을것 같아요 ㅋㅋㅋ

konempty reviewed (changes requested):
진짜 고생이 많으셨는데
쪼금만더 수정부탁드립니닷 ㅋㅋㅋㅋ...

Unagi-zoso committed:
ROFO-35 간략한 약관 조회 API 경로 변경

`summary` 는 없어도 오해할만한 다른 API 가 없어 생략해도 좋을 것 같다 판단했습니다.

Unagi-zoso committed:
ROFO-35 @JvmStatic 제거

큰 장점이 없다 판단해 제거합니다.

Unagi-zoso committed:
ROFO-35 간략한 약관 response 약관 사이즈 구하는 로직 추가

굳이 서비스에서 약관 사이즈들을 구하는 로직을 두지 않고 response dto의 생성자에서 구하도록 변경합니다.

Unagi-zoso committed:
ROFO-35 RequiredAndOptionalSizes 클래스 및 관련 메소드 제거

Unagi-zoso committed:
ROFO-35 checkRequiredTermsOrThrow 메소드 내부 subtract -> minus

Unagi-zoso committed:
ROFO-35 누락된 @Transactional 추가

Unagi-zoso committed:
ROFO-35 verify 시 횟수 명시 및 Mock 객체 초기화 부분 추가

Unagi-zoso committed:
ROFO-35 SummaryTermsResponse 잘못된 로직 수정

konempty commented on src/main/kotlin/kr/weit/roadyfoody/term/service/dto/DetailedTermResponse.kt:
```diff
@@ -0,0 +1,20 @@
+data class DetailedTermResponse(
+    val id: Long,
+    val title: String,
+    val isRequired: Boolean,
```
DB엔티티명은 required인데 DTO로는 isRequired라고 내려가는군요..!
요거 통일시켜도 좋을것 같아요!

Unagi-zoso replied:
아하 이건 민성님한테도 의견을 좀 여쭤봐서 결정했었던 부분이네요.
제 생각엔 팀장님이 피드백 주셨던 것처럼 코드상에선 Boolean 타입임을 잘 추적할 수 있어서 (IDE 지원도 있으며)
Flag 를 제외하는게 좋다 느꼈습니다. 대신 클라이언트에게 JSON 으로 내려보낼 땐 이게 Boolean 인지 확인할 수 있는 요소가
코드를 다룰 때보단 부족할 것 같다 생각해 좀 더 명확하게 나타낼 수 있는 이름을 취했습니다.
이 부분은 민성님한테도 좀 상의를 드렸고 조언해주셨던 부분을 살짝 반영해봤습니다.

konempty commented on src/main/resources/db/migration/V2__create_terms_and_user_agreed_terms.sql:
```diff
@@ -0,0 +1,42 @@
+ALTER TABLE user_agreed_terms
+    ADD CONSTRAINT user_agreed_terms_pk PRIMARY KEY (id);
+
+ALTER TABLE user_agreed_terms
+    ADD CONSTRAINT user_agreed_terms_user_id_term_id_uq UNIQUE (user_id, term_id);
+
+ALTER TABLE user_agreed_terms
+    ADD CONSTRAINT user_agreed_terms_user_id_fk foreign key (user_id) REFERENCES users (id);
```
한가지 재미있는 사실을 알려드리자면
복합키로 user_id, term_id로 잡혀있다면
user_id는 별도로 인덱스를 선언하지 않아도 인덱스가 잡힌효과가 발생해요 ㅋㅋㅋ
하지만 term_id은 user_id없이는 인덱스로 작동을 하지 않아 필요하다면 인덱스 선언이 필수예요
그래서 복합키를 잡을땐 순서에 유의를 해야합니다 ㅋㅋㅋㅋ

그래도 저는 명시적으로 User_id에 인덱스를 거는게 낫다고 생각합니다~!

Unagi-zoso replied:
와... 이런 세세한 부분은 잘 몰랐네요. 감사합니다!

konempty reviewed (approved):
정말 고생많으셨어요 👍 👍 
제가 너무 빡빡하게 리뷰를 드리지 않았나 싶긴한데 너무 잘 따라와주셔서 감사합니다!
리뷰는 읽어만 보시고 아 그렇구나 하고 넘어가도 되는 내용들입니다~!

Unagi-zoso commented:
> 빡하게 리뷰를 드리지 않았나 싶긴한데 너무 잘 따라와주셔서

피드백 너무 감사합니다! 
더 빡빡하게 주셔도 좋으니 부족한 부분 보이면 피드백 잘 부탁드리겠습니다!
